### DISCUSSION TITLE
DNS Check Error

### DISCUSSION THREAD

**Discussion by Ch4mpa9ne** [postalserver/postal#2325] [Help with using Postal]
## Describe the bug

When in DNS Setup for, After pressing the "Check my record" button, It always shows "There seems to be something wrong with your DNS records. Check below for information." 

I checked all dns resolution in every container, used `docker exec -it -u 0 postal-web-1 /bin/bash; apt update; apt install dnsutils.`

 And Whether inside or outside the container, the DNS resolution is always correct. I SWEAR IT!

But it always shows "something wrong with your DNS records balabala". I really don't know what else to try to debug. Logs has nothing wrong.

I have try the things below add to `/opt/postal/config/postal.yml`, but no help.

```
general:
  use_local_ns_for_domains: true
```

In addition, I am in an intranet environment, using a privately deployed DNS domain sever. But I'm sure that all the required dns fields have been resolved correctly, and I used dig to test repeatedly. 

## To Reproduce

1. Go to 'postal.example.com/org/mailorg/servers/mailorg/domain'
2. Click on 'Check muy record'
3. Scroll down to 'There seems to be something wrong with your DNS records. Check below for information.'

## Expected behaviour

Let it check it all correct.

## Screenshots

<img width="540" alt="image" src="https://user-images.githubusercontent.com/22672207/213429702-50cc2d5b-2df3-4a17-af92-a18bc738e22b.png">

## Environment details 

 - OS: Ubuntu 22.04
 - Browser Chrome
 - Version 2.1.2
 - Type Web

## Additional information/context

I can't think of anything else to provide. After I press the button, I check the postal_ worker_ 1' s log is normal, and there is even no new record. Tell me directly if you need anything.

**Comment by willpower232** [comment]
If the problem is in the UI then you need to check `postal logs web` to see what the issue is.

If you know how to use browser developer tools then you may be able to see what the error is in the Network tab but its more likely that the correct log will have the information in.

Regarding the problem, there is one setting which you might need to change https://github.com/postalserver/postal/blob/a7dd19baac8300f4d8ee89d0050479e08fdf9176/config/postal.defaults.yml#L17

**Reply by Ch4mpa9ne** [reply]
I'm sorry I wrote it wrong above, I tried to follow this answer before. In fact I tried false instead of true. but it still doesn't work

<img width="718" alt="image" src="https://user-images.githubusercontent.com/22672207/213473265-4416ff82-f340-4774-91b6-66cf15dc0745.png">

I think this should have nothing to do with web ui. I have also tried to observe the log of postal_web, but there seems to be no error. There is some logs in postal_ web when I click "Check my records are correct"

```
[10] [2023-01-19T14:55:26.881] INFO -- : Started GET "/org/mypostalmail/servers/mypostalmail/domains/a456fb32-85cb-4727-a04f-10f9fb4b027c/setup" for 192.168.100.178 at 2023-01-19 14:55:26 +0000
[10] [2023-01-19T14:55:26.882] INFO -- : Processing by DomainsController#setup as HTML
[10] [2023-01-19T14:55:26.882] INFO -- :   Parameters: {"org_permalink"=>"mypostalmail", "server_id"=>"mypostalmail", "id"=>"a456fb32-85cb-4727-a04f-10f9fb4b027c"}
[10] [2023-01-19T14:55:26.893] INFO -- :   Rendering domains/setup.html.haml within layouts/application
[10] [2023-01-19T14:55:26.896] INFO -- :   Rendered servers/_sidebar.html.haml (2.7ms)
[10] [2023-01-19T14:55:26.902] INFO -- :   Rendered servers/_header.html.haml (6.1ms)
[10] [2023-01-19T14:55:26.903] INFO -- :   Rendered domains/_nav.html.haml (0.1ms)
[10] [2023-01-19T14:55:26.903] INFO -- :   Rendered domains/setup.html.haml within layouts/application (10.0ms)
[10] [2023-01-19T14:55:26.906] INFO -- : Completed 200 OK in 23ms (Views: 10.7ms | ActiveRecord: 6.2ms)
192.168.100.178 - - [19/Jan/2023:14:55:26 +0000] "GET /org/mypostalmail/servers/mypostalmail/domains/a456fb32-85cb-4727-a04f-10f9fb4b027c/setup HTTP/1.1" 200 - 0.0264
```

**Reply by willpower232** [reply]
it doesn't look like Postal is having any issues, just that it doesn't think you have updated your DNS.

Have you tried using https://whatsmydns.net to verify what your DNS records are around the world?

**Reply by Ch4mpa9ne** [reply]
> it doesn't look like Postal is having any issues, just that it doesn't think you have updated your DNS.
> 
> Have you tried using https://whatsmydns.net to verify what your DNS records are around the world?

Does the Postal service always check DNS resolution from the public network? In fact, I am deploying the Postal service on the intranet  as a intranet's email server which is Isolated from the Internet. DNS only takes effect on the intranet and cannot be resolved on the public network.

I confirmed again that the container can obtain the DNS resolution (only from the internal network) related to the mail that I configured.

**Reply by willpower232** [reply]
fundamentally, Postal will allow you to send the messages if the domain is listed in the UI, regardless of how many records are apparently not correct.

You can try sending the messages and see how you get on, the DNS records are mostly to guarantee that the message will be received by the intended users, I'm not sure how that works from an intranet point of view but presumably you could whitelist the sending email addresses you're using if it comes to it.

**Comment by bganderson** [comment]
Just had a similar problem to this and it was driving me nuts. Considering @Ch4mpa9ne was running this with an internal resolver, it seems likely that this is the same problem. 

TL;DR **Make sure your domains zone has `NS` records configured that point to your authoritative DNS server.**

### Details

Postal verifies the records by first getting the nameservers (not from whois) from DNS `NS` entries for the domain:

https://github.com/postalserver/postal/blob/fd3c7ccdf6dc4ee0a76c9523cbd735159e4b8000/app/lib/dns_resolver.rb#L137

If you want to test to see if this is your issue, you can `docker exec -it <CONTAINER_ID> /bin/sh` and `rails console` and:
```
d = Domain.new(name: "yourdomain.tld")
d.resolver
```
That should return a resolver with the set nameservers looking something like this normally:
`@nameservers=["1.2.3.4", "5.6.7.8"]>`
But in my case I was getting:
`@nameservers=[]>`

After adding `NS` records and waiting for propagation, everything was happy. 

### Suggestion

Display an error if `effective_ns` returns `[]` saying something like "Unable to determine nameservers. Check NS records."

The error that gets displayed, "No SPF record exists for this domain",  had me spending a bunch of time troubleshooting container DNS resolution.
